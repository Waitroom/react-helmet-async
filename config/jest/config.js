const path = require('path');

module.exports = {
  collectCoverage: true,
  collectCoverageFrom: [
    'src/**/*.{js,jsx,ts,tsx}',
  ],
  coverageDirectory: '__coverage__',
  rootDir: '../..',
  testMatch: ['**/__tests__/**/*.(j|t)s?(x)'],
  testPathIgnorePatterns: [
    '/__assets__/',
    '/node_modules/',
  ],
  testEnvironmentOptions: {
    url: 'http://localhost',
  },
  transform: {
    '\\.((j|t)sx?)$': 'babel-jest',
  },
  setupFilesAfterEnv: [
    '<rootDir>/config/jest/setup.js',
  ],
};