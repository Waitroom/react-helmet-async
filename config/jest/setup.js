// Polyfills requestAnimationFrame() in the test environment.
import 'raf/polyfill';
