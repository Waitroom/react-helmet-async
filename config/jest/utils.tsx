import { type ReactNode, act, StrictMode } from 'react';
import { createRoot } from 'react-dom/client';
import type { Root } from 'react-dom/client';

import Provider from '../../src/Provider';
import type { HelmetDataContext } from '../../src/HelmetData';

let root: Root | null = null;

export const unmount = () => {
  act(() => {
    root?.unmount();
    root = null;
  });
};

export const render = (node: ReactNode, context = {}) => {
  if (!root) {
    const elem = document.getElementById('mount');
    if (!elem) throw Error('Internal error');
    root = createRoot(elem);
  }

  act(() => {
    root?.render(
      <StrictMode>
        <Provider context={context}>{node}</Provider>
      </StrictMode>,
    );
  });
};

export const renderContext = (node: ReactNode) => {
  const context = {} as Partial<HelmetDataContext>;
  render(node, context);
  return context.helmet;
};

export const isArray = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  asymmetricMatch: (actual: any) => Array.isArray(actual),
};
