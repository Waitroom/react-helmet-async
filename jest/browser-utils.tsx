import { type ReactNode, act, StrictMode } from 'react';
import { createRoot, type Root } from 'react-dom/client';

import Provider from '../src/Provider';
import type { HelmetDataContext, HelmetServerState } from '../src/types';

let root: Root | null = null;

export const unmount = (): void => {
  act(() => {
    root?.unmount();
    root = null;
  });
};

export const renderClient = (node: ReactNode, context = {}): void => {
  if (!root) {
    const elem = document.getElementById('mount');
    if (!elem) throw Error('Internal error');
    root = createRoot(elem);
  }

  act(() => {
    root?.render(
      <StrictMode>
        <Provider context={context}>{node}</Provider>
      </StrictMode>,
    );
  });
};

export const renderContextClient = (
  node: ReactNode,
): HelmetServerState | undefined => {
  const context: HelmetDataContext = {};
  renderClient(node, context);
  return context.helmet;
};

// TODO: Get rid of this method.
export const isArray = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  asymmetricMatch: (actual: any): boolean => Array.isArray(actual),
};
