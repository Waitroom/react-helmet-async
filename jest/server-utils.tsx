// TODO: Remove client-side utils from this module, they belong to browser-utils
// module.

import { type ReactNode, StrictMode } from 'react';

import { renderToStaticMarkup } from 'react-dom/server';

import Provider from '../src/Provider';
import type { HelmetDataContext } from '../src/types';

/**
 * Renders the given `node` within the provided `context` into HTML markup,
 * using server-side rendering API.
 */
export function renderServer(node: ReactNode, context = {}) {
  return renderToStaticMarkup(
    <StrictMode>
      <Provider context={context}>{node}</Provider>
    </StrictMode>,
  );
}

export function renderContextServer(node: ReactNode) {
  const context: HelmetDataContext = {};
  renderServer(node, context);
  return context.helmet;
}

// TODO: Get rid of this method.
export const isArray = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  asymmetricMatch: (actual: any) => Array.isArray(actual),
};
