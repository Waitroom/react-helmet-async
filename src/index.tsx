export * from './types';

export { default as Helmet } from './Helmet';
export { default as HelmetProvider } from './Provider';
