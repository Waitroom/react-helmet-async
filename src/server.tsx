import { type Key, type ReactNode, createElement } from 'react';

import {
  HELMET_DATA_ATTRIBUTE,
  TAG_NAMES,
  REACT_TAG_MAP,
  TAG_PROPERTIES,
  SEO_PRIORITY_TAGS,
} from './constants';

import { flattenArray, prioritizer } from './utils';

import type {
  BodyProps,
  HelmetDatum,
  HelmetHTMLBodyDatum,
  HelmetHTMLElementDatum,
  HelmetServerState,
  HtmlProps,
  MappedServerState,
} from './types';

const SELF_CLOSING_TAGS: string[] = [TAG_NAMES.NOSCRIPT, TAG_NAMES.SCRIPT, TAG_NAMES.STYLE];

const encodeSpecialCharacters = (str: string, encode = true) => {
  if (encode === false) {
    return String(str);
  }

  return String(str)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;');
};

type Attributes = Record<string, string | null | number | boolean | undefined>;

function generateElementAttributesAsString(attrs: Attributes): string {
  let res: string = '';

  for (const [name, value] of Object.entries(attrs)) {
    const neu = value === undefined ? name : `${name}="${value}"`;
    if (neu && res) res += ' ';
    res += neu;
  }

  return res;
}

const generateTitleAsString = (
  title: string,
  attrs: Attributes,
  encode: boolean,
) => {
  let attrsStr = generateElementAttributesAsString(attrs);
  if (attrsStr) attrsStr = ` ${attrsStr}`;

  const flattenedTitle = flattenArray(title);

  return `<title ${HELMET_DATA_ATTRIBUTE}="true"${attrsStr}>${
    encodeSpecialCharacters(flattenedTitle, encode)
  }</title>`;
};

// TODO: Refactor!
const generateTagsAsString = (
  type: string,
  tags: HTMLElement[],
  encode = true,
) => tags.reduce((str, t) => {
  const tag = t as unknown as Attributes;
  const attributeHtml = Object.keys(tag)
    .filter(
      (attribute) => !(attribute === TAG_PROPERTIES.INNER_HTML as string || attribute === TAG_PROPERTIES.CSS_TEXT as string),
    )
    .reduce((string, attribute) => {
      const attr = typeof tag[attribute] === 'undefined'
        ? attribute
        : `${attribute}="${encodeSpecialCharacters(tag[attribute] as string, encode)}"`;
      return string ? `${string} ${attr}` : attr;
    }, '');

  const tagContent = tag.innerHTML ?? tag.cssText ?? '';

  const isSelfClosing = !SELF_CLOSING_TAGS.includes(type);

  return `${str}<${type} ${HELMET_DATA_ATTRIBUTE}="true" ${attributeHtml}${
    isSelfClosing ? '/>' : `>${tagContent}</${type}>`
  }`;
}, '');

/**
 * Given a map of element attribute names & values it returns the corresponding
 * map of element properties & values (i.e. replacing some attribute names by
 * their corresponding property names).
 */
function mapElementAttributesToProps(
  attributes: Attributes | HTMLElement,
  ops: { addHelmetDataAttr?: boolean; addKey?: Key } = {},
): Record<string, unknown> {
  const res: Record<string, unknown> = {};
  if (ops.addHelmetDataAttr) res[HELMET_DATA_ATTRIBUTE] = true;
  if (ops.addKey !== undefined) res.key = ops.addKey;
  for (const [attrName, value] of Object.entries(attributes)) {
    const propName = REACT_TAG_MAP[attrName] ?? attrName;

    switch (propName) {
      // cssText and innerHTML props get a special treatment to avoid that React
      // escapes their values.
      case 'cssText':
      case 'innerHTML':
        res.dangerouslySetInnerHTML = { __html: value as unknown };
        break;
      default:
        res[propName] = value;
    }
  }
  return res;
}

function renderTitle(title: string, attrs: Attributes): ReactNode {
  // NOTE: Rendered as array to match legacy behavior.
  return [
    <title
      key={title}
      {...mapElementAttributesToProps(attrs, { addHelmetDataAttr: true })}
    >
      {title}
    </title>,
  ];
}

function renderElement(type: string, attrs: HTMLElement, key?: Key): ReactNode {
  return createElement(type, mapElementAttributesToProps(attrs, {
    addHelmetDataAttr: true,
    addKey: key,
  }));
}

function renderElements(type: string, attrs: HTMLElement[]): ReactNode[] {
  const res: ReactNode[] = [];
  for (let i = 0; i < attrs.length; ++i) {
    res.push(renderElement(type, attrs[i]!, i));
  }
  return res;
}

function newHelmetDatumForAttrs(attrs: BodyProps): HelmetHTMLBodyDatum;
function newHelmetDatumForAttrs(attrs: HtmlProps): HelmetHTMLElementDatum;

function newHelmetDatumForAttrs(
  attrs: BodyProps | HtmlProps,
): HelmetHTMLBodyDatum | HelmetHTMLElementDatum {
  return {
    toComponent: () => mapElementAttributesToProps(attrs),
    toString: () => generateElementAttributesAsString(attrs),
  };
}

function newHelmetDatumForTags(
  type: string,
  attrs: HTMLElement[],
  encode: boolean,
): HelmetDatum {
  return {
    toComponent: () => renderElements(type, attrs),
    toString: () => generateTagsAsString(type, attrs, encode),
  };
}

function newHelmetDatumForTitle(
  title: string,
  attrs: Attributes = {},
  encode: boolean,
): HelmetDatum {
  return {
    toComponent: () => renderTitle(title, attrs),
    toString: () => generateTitleAsString(title, attrs, encode),
  };
}

const getPriorityMethods = ({
  metaTags,
  linkTags,
  scriptTags,
  encode,
}: MappedServerState) => {
  const meta = prioritizer(metaTags, SEO_PRIORITY_TAGS.meta);
  const link = prioritizer(linkTags, SEO_PRIORITY_TAGS.link);
  const script = prioritizer(scriptTags, SEO_PRIORITY_TAGS.script);

  // need to have toComponent() and toString()
  const priorityMethods = {
    toComponent: () => [
      ...renderElements(TAG_NAMES.META, meta.priority),
      ...renderElements(TAG_NAMES.LINK, link.priority),
      ...renderElements(TAG_NAMES.SCRIPT, script.priority),
    ],

    // generate all the tags as strings and concatenate them
    toString: () => `${
      generateTagsAsString(TAG_NAMES.META, meta.priority, encode)} ${
      generateTagsAsString(TAG_NAMES.LINK, link.priority, encode)} ${
      generateTagsAsString(TAG_NAMES.SCRIPT, script.priority, encode)}`,
  };

  return {
    priorityMethods,
    metaTags: meta.default as HTMLMetaElement[],
    linkTags: link.default as HTMLLinkElement[],
    scriptTags: script.default as HTMLScriptElement[],
  };
};

function mapStateOnServer(props: MappedServerState): HelmetServerState {
  const {
    baseTag,
    bodyAttributes,
    encode = true,
    htmlAttributes,
    noscriptTags,
    prioritizeSeoTags,
    styleTags,
    title = '',
    titleAttributes,
  } = props;
  let { linkTags, metaTags, scriptTags } = props;
  let priorityMethods: HelmetDatum = {
    toComponent: () => null,
    toString: () => '',
  };
  if (prioritizeSeoTags) {
    ({
      priorityMethods,
      linkTags,
      metaTags,
      scriptTags,
    } = getPriorityMethods(props));
  }
  return {
    priority: priorityMethods,
    base: newHelmetDatumForTags(TAG_NAMES.BASE, baseTag, encode),
    bodyAttributes: newHelmetDatumForAttrs(bodyAttributes ?? {}),
    htmlAttributes: newHelmetDatumForAttrs(htmlAttributes ?? {}),
    link: newHelmetDatumForTags(TAG_NAMES.LINK, linkTags, encode),
    meta: newHelmetDatumForTags(TAG_NAMES.META, metaTags, encode),
    noscript: newHelmetDatumForTags(TAG_NAMES.NOSCRIPT, noscriptTags, encode),
    script: newHelmetDatumForTags(TAG_NAMES.SCRIPT, scriptTags, encode),
    style: newHelmetDatumForTags(TAG_NAMES.STYLE, styleTags, encode),
    title: newHelmetDatumForTitle(title, titleAttributes, encode),
  };
}

export default mapStateOnServer;
