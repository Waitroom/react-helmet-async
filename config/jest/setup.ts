// Polyfills requestAnimationFrame() in the test environment.
import 'raf/polyfill';

import '@testing-library/jest-dom';

/*
import '@testing-library/jest-dom';

import { clearInstances } from '../src/HelmetData';

import { unmount } from './utils';

// @ts-expect-error "pre-existing"
globalThis.IS_REACT_ACT_ENVIRONMENT = true;

let headElement: HTMLHeadElement;

beforeEach(() => {
  headElement ||= document.head || document.querySelector('head');

  headElement.innerHTML = '';
  document.body.innerHTML = '<div id="mount"></div>';
});

afterEach(() => {
  unmount();

  clearInstances();
});

*/
