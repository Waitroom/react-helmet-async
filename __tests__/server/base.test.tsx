import { renderToStaticMarkup } from 'react-dom/server';

import { Helmet } from '../../src';
import { renderContextServer, isArray } from '../../jest/server-utils';

describe('server', () => {
  describe('API', () => {
    it('renders base tag as React component', () => {
      const head = renderContextServer(
        <Helmet base={{ href: 'http://localhost/', target: '_blank' }} />,
      );

      expect(head?.base).toBeDefined();
      expect(head?.base.toComponent).toBeDefined();

      const baseComponent = head?.base.toComponent();

      expect(baseComponent).toStrictEqual(isArray);
      expect(baseComponent).toHaveLength(1);

      (baseComponent as unknown as Element[]).forEach((base: Element) => {
        expect(base).toStrictEqual(expect.objectContaining({ type: 'base' }));
      });

      const markup = renderToStaticMarkup(baseComponent);

      expect(markup).toMatchSnapshot();
    });

    it('renders base tags as string', () => {
      const head = renderContextServer(
        <Helmet base={{ href: 'http://localhost/', target: '_blank' }} />,
      );
      expect(head?.base).toBeDefined();
      expect(head?.base.toString).toBeDefined();
      expect(head?.base.toString()).toMatchSnapshot();
    });
  });

  describe('Declarative API', () => {
    it('renders base tag as React component', () => {
      const head = renderContextServer(
        <Helmet>
          <base href="http://localhost/" target="_blank" />
        </Helmet>,
      );

      expect(head?.base).toBeDefined();
      expect(head!.base.toComponent).toBeDefined();

      const baseComponent = head?.base.toComponent();

      expect(baseComponent).toStrictEqual(isArray);
      expect(baseComponent).toHaveLength(1);

      (baseComponent as unknown as Element[]).forEach((base: Element) => {
        expect(base).toStrictEqual(expect.objectContaining({ type: 'base' }));
      });

      const markup = renderToStaticMarkup(baseComponent);

      expect(markup).toMatchSnapshot();
    });

    it('renders base tags as string', () => {
      const head = renderContextServer(
        <Helmet>
          <base href="http://localhost/" target="_blank" />
        </Helmet>,
      );

      expect(head?.base).toBeDefined();
      expect(head!.base.toString).toBeDefined();
      expect(head?.base.toString()).toMatchSnapshot();
    });
  });
});
