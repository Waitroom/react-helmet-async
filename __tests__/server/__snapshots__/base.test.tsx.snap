// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server API renders base tag as React component 1`] = `"<base data-rh="true" href="http://localhost/" target="_blank"/>"`;

exports[`server API renders base tags as string 1`] = `"<base data-rh="true" href="http://localhost/" target="_blank"/>"`;

exports[`server Declarative API renders base tag as React component 1`] = `"<base data-rh="true" href="http://localhost/" target="_blank"/>"`;

exports[`server Declarative API renders base tags as string 1`] = `"<base data-rh="true" href="http://localhost/" target="_blank"/>"`;
